sqlitecluster/SQLite.cpp

```diff
@@ -256,79 +308,139 @@ bool SQLite::write(const string& query) {
     return true;
 }
 
-// --------------------------------------------------------------------------
-void SQLite::commit() {
+int SQLite::commit() {
     SASSERT(_insideTransaction);
     SASSERT(!_uncommittedHash.empty()); // Must prepare first
+
+    int result = 0;
     // Do we need to truncate as we go?
     bool truncating = false;
     if (_journalSize + 1 > _maxJournalSize) {
         // Delete the oldest entry
         truncating = true;
         uint64_t before = STimeNow();
-        SASSERT(SQuery(_db, "Deleting oldest row",
-                       "DELETE FROM journal WHERE id=" + SToStr(_commitCount - _journalSize + 1)));
+        string query = "DELETE FROM" + _journalName + " WHERE id = MIN(id);";
+        SASSERT(!SQuery(_db, "Deleting oldest row", query));
         _writeElapsed += STimeNow() - before;
     }
 
     // Make sure one is ready to commit
     SDEBUG("Committing transaction");
     uint64_t before = STimeNow();
-    SASSERT(SQuery(_db, "committing db transaction", "COMMIT"));
-    _commitElapsed += STimeNow() - before;
+    result = SQuery(_db, "committing db transaction", "COMMIT");
 
-    // Successful commit
-    ++_commitCount;
-    _journalSize += !truncating; // Only increase if we didn't truncate by a row
-    _committedHash = _uncommittedHash;
-    _insideTransaction = false;
-    _uncommittedHash.clear();
-    _uncommittedQuery.clear();
+    // If there were conflicting commits, will return SQLITE_BUSY_SNAPSHOT
+    SASSERT(result == SQLITE_OK || result == SQLITE_BUSY_SNAPSHOT);
+
+    if (result == SQLITE_OK) {
+        _commitElapsed += STimeNow() - before;
+        // Successful commit
+        _journalSize += !truncating; // Only increase if we didn't truncate by a row
+
+        _committedtransactionIDs.insert(_uncommittedID);
+        {
+            // Update atomically.
+            lock_guard<recursive_mutex> lock(_hashLock);
+            _lastCommittedHash = _uncommittedHash;
+        }
+
+        // Ok, someone else can commit now.
+        // Incrementing the commit count.
+        _commitCount++;
+        SINFO("Commit successful (" << _commitCount.load() << "), releasing commitLock.");
+
+        _insideTransaction = false;
+        _uncommittedTransaction = false;
+        _uncommittedHash.clear();
+        _uncommittedQuery.clear();
+        commitLock.unlock();
+    } else {
+        SINFO("Commit failed, waiting for rollback.");
+    }
+
+    // if we got SQLITE_BUSY_SNAPSHOT, then we're *still* holding commitLock, and it will need to be unlocked by
+    // calling rollback().
+    return result;
+}
+
+map<uint64_t, pair<string,string>> SQLite::getCommittedTransactions() {
+    SQLITE_COMMIT_AUTOLOCK;
+
+    map<uint64_t, pair<string,string>> result;
+    if (_committedtransactionIDs.empty()) {
+        return result;
+    }
+
+    for (uint64_t key : _committedtransactionIDs) {
+        result[key] = move(_inFlightTransactions.at(key));
```

Comment:
std::move essentially casts its argument to a temporary variable. C++11 introduces move semantics, which let you avoid a lot of unnecessary copying of objects. Imagine a hypothetical class:

```
class Bitmap {
  public:
    Bitmap(int width * int height);
    ~Bitmap();
  private:
    void* imageData;
}

Bitmap::Bitmap(int width, int height) {
    imageData = malloc(width * height);
}

Bitmap::~Bitmap() {
    free(imageData);
}
```

Now, say you want to store a bunch of these in an  STL list:
```
list<Bitmap> pictures;
for (int i = 0; i < lots; i++) {
    Bitmap temp(1000, 1000); // 1MB square image.
    pictures.push_back(temp);
}
```

What this will do, is for each image, allocate a megabyte of image data, then push that onto `pictures` by calling the copy constructor, which will allocate *another* megabyte memory, copy the memory from the first object, and then call the destructor for the first object, freeing that memory. This is clearly wasteful.

So in C++11, they've added the notion of an `r-value reference` which is essentially a reference to a temporary object. Imagine the following code in place of what we used above:

```
for (int i = 0; i < lots; i++) {
    pictures.push_back(Bitmap(1000, 1000)); // 1MB square image.
}
```

The bitmap we're pushing onto `pictures` here is a temporary object, so the compiler can call the move constructor instead of the copy constructor. If we implement a move constructor for `Bitmap` it might look like this:

```
// The double "&&" is an r-value reference.
Bitmap::Bitmap(Bitmap && from) {
    imageData = from.imageData;
    from.imageData = 0;
}
```

Now, by calling the move constructor instead of the copy constructor, we do the allocation only once, and the `free` on the temporary object becomes a no-op, which is much more efficient. We know this is safe to do because `from` is temporary, and thus nothing will be able to try and access it when we're done with it.

So, what does `move` do: it casts it's argument to a temporary. Let's look at the first `for loop again, but we'll add a `move`:
```
for (int i = 0; i < lots; i++) {
    Bitmap temp(1000, 1000); // 1MB square image.
    pictures.push_back(move(temp));
}
```

This does the same thing as the version where we pass the temporary `Bitmap`, because we still get to call the move constructor to add the bitmap to our list. Importantly, this leaves `temp` in a broken and indeterminate state after the move, but it's going out of scope on the next line anyway. This is how we use `move` here - we move this item out of `_inFlightTransactions`, but the very next thing we do is delete the item anyway, so it doesn't matter that we may have left it in a broken state.